### ISSUE TITLE
Pipeline Request till access token is set in Interceptor

### ISSUE THREAD

**Issue by Yrajaram112** [axios/axios#5895]
### Is your feature request related to a problem? Please describe.

When multiple api calls are made simultaneously for then interceptor currently is keeping all the response in queue and then recall then original request.

Instead , I want something like pipelined request where if one api call throws 401 or 403 then keep other request in hold in request interceptor not like current practise of keeping failed queue.

Something like this have been implemented in Angular using RxJS.

### Describe the solution you'd like

Make some flag to denote that the first request has returned 401 or 403 response so instead of calling all other api and then all other returning 403 status and then the current practise of recalling the failed api queue 
Keep other request in request pipeline if one api fails and is refreshing.
If refreshtoken set the new token then only process the request.

### Additional context/Screenshots

```bash
I found similar implementation using RxJs in angular where they pipeline the request when one response fail


import {throwError as observableThrowError,  Observable ,  BehaviorSubject } from 'rxjs';

import {take, filter, catchError, switchMap, finalize} from 'rxjs/operators';
import { Injectable, Injector } from "@angular/core";
import { HttpInterceptor, HttpRequest, HttpHandler, HttpSentEvent, HttpHeaderResponse, HttpProgressEvent, HttpResponse, HttpUserEvent, HttpErrorResponse } from "@angular/common/http";

import { AuthService } from "./auth.service";

@Injectable()
export class RequestInterceptorService implements HttpInterceptor {

    isRefreshingToken: boolean = false;
    tokenSubject: BehaviorSubject<string> = new BehaviorSubject<string>(null);

    constructor(private injector: Injector) {}

    addToken(req: HttpRequest<any>, token: string): HttpRequest<any> {
        return req.clone({ setHeaders: { Authorization: 'Bearer ' + token }})
    }

    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpSentEvent | HttpHeaderResponse | HttpProgressEvent | HttpResponse<any> | HttpUserEvent<any>> {
        const authService = this.injector.get(AuthService);

        return next.handle(this.addToken(req, authService.getAuthToken())).pipe(
            catchError(error => {
                if (error instanceof HttpErrorResponse) {
                    switch ((<HttpErrorResponse>error).status) {
                        case 400:
                            return this.handle400Error(error);
                        case 401:
                            return this.handle401Error(req, next);
                        default:
                            return observableThrowError(error);
                    }
                } else {
                    return observableThrowError(error);
                }
            }));
    }

    handle400Error(error) {
        if (error && error.status === 400 && error.error && error.error.error === 'invalid_grant') {
            // If we get a 400 and the error message is 'invalid_grant', the token is no longer valid so logout.
            return this.logoutUser();
        }

        return observableThrowError(error);
    }

    handle401Error(req: HttpRequest<any>, next: HttpHandler) {
        if (!this.isRefreshingToken) {
            this.isRefreshingToken = true;

            // Reset here so that the following requests wait until the token
            // comes back from the refreshToken call.
            this.tokenSubject.next(null);

            const authService = this.injector.get(AuthService);
            
            return authService.refreshToken().pipe(
                switchMap((newToken: string) => {
                    if (newToken) {
                        this.tokenSubject.next(newToken);
                        return next.handle(this.addToken(this.getNewRequest(req), newToken));
                    }

                    // If we don't get a new token, we are in trouble so logout.
                    return this.logoutUser();
                }),
                catchError(error => {
                    // If there is an exception calling 'refreshToken', bad news so logout.
                    return this.logoutUser();
                }),
                finalize(() => {
                    this.isRefreshingToken = false;
                }),);
        } else {
            return this.tokenSubject.pipe(
                filter(token => token != null),
                take(1),
                switchMap(token => {
                    return next.handle(this.addToken(this.getNewRequest(req), token));
                }),);
        }
    }

    /*
        This method is only here so the example works.
        Do not include in your code, just use 'req' instead of 'this.getNewRequest(req)'.
    */
    getNewRequest(req: HttpRequest<any>): HttpRequest<any> {
        if (req.url.indexOf('getData') > 0) {
            return new HttpRequest('GET', 'http://private-4002d-testerrorresponses.apiary-mock.com/getData');
        }

        return new HttpRequest('GET', 'http://private-4002d-testerrorresponses.apiary-mock.com/getLookup');
    }

    logoutUser() {
        // Route to the login page (implementation up to you)

        return observableThrowError("");
    }
}
```

**Comment by chandra-pro** [comment]
@Yrajaram112 Can you assign me this issue?
I would like to work on this project

**Comment by jasonsaayman** [comment]
Thanks for writing this up, and sorry it sat for so long.

The problem is real. When several requests go out together and the token has expired, every one of them burns a round trip and comes back 401 before anything gets refreshed. Holding the queued ones until the refresh lands is the right instinct.

I am not going to add it to core though, and I want to be straight about why. A gate like this needs to know things axios does not: which requests are authenticated, which one is the refresh call itself (that one must never wait on the lock, or you deadlock), and what should happen to the queue when the refresh fails. All of that is application state. What axios owes you is the hooks to build it, and those already exist.

The promise-based equivalent of the RxJS pipeline you linked is a shared promise that the request interceptor awaits:

```js
let refreshing = null;

api.interceptors.request.use(async (config) => {
  if (refreshing && !config.skipAuthRefresh) {
    await refreshing;
  }
  config.headers.Authorization = 'Bearer ' + getToken();
  return config;
});

api.interceptors.response.use(null, async (error) => {
  const config = error.config;

  if (!error.response || error.response.status !== 401 || config._retried) {
    return Promise.reject(error);
  }

  config._retried = true;
  refreshing = refreshing || refreshToken().finally(() => { refreshing = null; });

  await refreshing;
  return api(config);
});
```

One thing worth knowing, because it is the part people get wrong: the request interceptor only holds requests that have not been dispatched yet. Anything already on the wire when the first 401 arrives will still come back 401, so you need the reactive half as well. The two halves together are the answer, not either one alone.

That combination, including the case where a proactive refresh timer and a reactive 401 handler race each other over a rotating refresh token, is being written up as a documented recipe in #10701. That covers everything asked for here and more, so I am closing this in favour of it. Worth subscribing there.

@chandra-pro if you are still interested, #10701 is the place to help. The docs work there is open and would be genuinely useful.
